train/NEAT/evolve.py

```diff
@@ -0,0 +1,157 @@
+#  GLOBALS
+# -----------------------------------------------------------------------------
+args = None
+env = []
+
+for i in range(0, 32):
+    env.append(gym.make('ppaquette/SuperMarioBros-{:d}-{:d}-Tiles-v0'.format(int((i/4)+1), int((i%4)+1))))
+
+# -----------------------------------------------------------------------------
+#  NEAT TRAINING
+# -----------------------------------------------------------------------------
+
+# Creates valid button presses to pass to gym
+def clean_outputs(outputs):
+    return [1 if sigmoid(b) > 0.5 else 0 for b in outputs]
+
+# Simple sigmoid function. 
+def sigmoid(x):
+    return (1 / (1 + np.exp(-x)))
+
+#TODO: Need a better fitness function.
+def calculate_fitness(info):
+    return info['distance']
+
+def eval_genome(genome, config):
+
+    net = neat.nn.FeedForwardNetwork.create(genome, config)
+    stuckMax = 600
+    
+    for i in range(0, 32):
+        observation = env[i].reset()
+        done = False
+        stuck = 0
+        while not done:
+            # Get move from NN
+            outputs = clean_outputs(net.activate(observation.flatten()))
+
+            # Make move
+            observation, reward, done, info = env[i].step(outputs)
+
+            # Check if Mario is progressing in level
+            stuck += 1 if reward <= 0 else 0
+
+            #TODO: Needs improvment, need to disable at end of level and when in a pipe.
+            if (stuck > stuckMax):
+                env[i].close()
+                return calculate_fitness(info)
+
+        if info['life'] == 0:
+            break
+
+        env[i].close()
+    return calculate_fitness(info)
+
+def run(config):
+    if args.checkpoint:
+        pop = neat.Checkpointer.restore_checkpoint(args.checkpoint)
+    else:
+        pop = neat.Population(config)
+
+    # Checkpoint every 5 generations or 10 min.
+    pop.add_reporter(neat.Checkpointer(5, 600))
+
+    pop.add_reporter(neat.StdOutReporter(True))
+    stats = neat.StatisticsReporter()
+    pop.add_reporter(stats)
+
+    pe = neat.ParallelEvaluator(args.numCores, eval_genome)
+    
+    # We could use this to set a limit of generations... But i dont think we need too... 
+    #for x in range(args.generations if args.generations > 0 else sys.maxsize): 
+    while True:
+        winner = pop.run(pe.evaluate, 5)
+        
+        if args.vineLogging:
+            #TODO: Fix Vine Logging, everything needed should be in stats. 
+            visualise.save_statistics(stats)
+        else:
+            visualise.plot_stats(stats, ylog=False, view=False)
+            visualise.plot_species(stats, view=False)
+
+        # Save the best Genome from the last 5 gens.
+        with open('Best-{}.pkl'.format(len(stats.most_fit_genomes)), 'wb') as output:
+            pickle.dump(winner, output, 1)
+
+def playBest(config, bestFile):
+    #TODO: Set to normal...
+    # Error - The mode "normal" is not supported. Supported options are "algo" or "human"
```

Comment:
How come this doesn't work?

Reply:
Don't know, I didn't really look into it, i just assumed that Jasons changes haven't been put into develop yet.